src/Microsoft.DotNet.Build.CloudTestTasks/SendToEventHub.cs

```diff
@@ -1,53 +1,141 @@
+    using System.IO;
+    using System.Net;
+    using System.Security.Cryptography;
+    using System.Text;
+    using System.Threading.Tasks;
+
+    using Microsoft.Build.Framework;
+
+    using Task = Microsoft.Build.Utilities.Task;
+
+    #endregion
+
+    /// <summary>
+    ///     The send to event hub.
+    /// </summary>
+    public class SendToEventHub : Task
     {
         /// <summary>
-        /// The Service Bus connection string.
+        ///     The Service Bus connection string.
         /// </summary>
         [Required]
         public string ConnectionString { get; set; }
 
         /// <summary>
-        /// The path to the Event Hub.
+        ///     The event data used to form the body stream.
+        /// </summary>
+        [Required]
+        public string EventData { get; set; }
+
+        /// <summary>
+        ///     The path to the Event Hub.
         /// </summary>
         [Required]
         public string EventHubPath { get; set; }
 
         /// <summary>
-        /// The event data used to form the body stream.
+        ///     Gets or sets the event hub shared access key.
         /// </summary>
         [Required]
-        public string EventData { get; set; }
+        public string EventHubSharedAccessKey { get; set; }
 
         /// <summary>
-        /// The partition key for the event.
+        ///     The partition key for the event.
         /// </summary>
         public string PartitionKey { get; set; }
 
+        /// <summary>
+        ///     Gets or sets the content bytes.
+        /// </summary>
+        internal byte[] contentBytes { get; set; }
+
+        /// <summary>
+        ///     The execute.
+        /// </summary>
+        /// <returns>
+        ///     The <see cref="bool" />.
+        /// </returns>
         public override bool Execute()
         {
-            using (var streamReader = new StreamReader(EventData))
+            return this.ExecuteAsync().GetAwaiter().GetResult();
+        }
+
+        public async Task<bool> ExecuteAsync()
+        {
+            using (var fs = new FileStream(this.EventData, FileMode.Open, FileAccess.Read))
             {
-                EventHubClient client = EventHubClient.CreateFromConnectionString(ConnectionString, EventHubPath);
-                client.Send(new EventData(Encoding.UTF8.GetBytes(streamReader.ReadToEnd()))
+                using (var streamReader = new StreamReader(fs))
                 {
-                    PartitionKey = PartitionKey
-                });
-                Log.LogMessage(MessageImportance.Normal, "Successfully sent notification to event hub path {0}.", EventHubPath);
+                    var request = (HttpWebRequest)WebRequest.Create(this.EventHubPath);
+                    request.Method = "POST";
+                    request.ContentType = "application/atom+xml;type=entry;charset=utf-8";
+
+                    var expiry = (int)DateTime.UtcNow.AddMinutes(20).Subtract(new DateTime(1970, 1, 1)).TotalSeconds;
```

Comment:
This is quite a bit of date math, a little comment in english about what it's trying to say (which is, I think "The api expects times in 'number of seconds since 1970', and we want the upload good for 20 minutes from now)